lib/rpush/daemon/apnsp8/delivery.rb

```diff
@@ -14,11 +14,11 @@ def initialize(app, http2_client, token_provider, batch)
           @batch = batch
           @first_push = true
           @token_provider = token_provider
-        end
 
-        def perform
           @client.on(:error) { |err| mark_batch_retryable(Time.now + 10.seconds, err) }
```

Comment:
I saw that this is also the case for regular `apns2`, as in, it's called in `perform`. Should we also change it there?

My question would also be, have you verified this works as intended if it's called in `initialize`? This construct (`@client.on`) is only used in `apns2` and `apnsp8` and nowhere else in rpush which makes me think it was copied over from `apns` during the `apnsp8` implementation, but I'm not sure exactly how it came to be.


https://github.com/rpush/rpush/blob/a8e7798de16c645b0185fb81950fdac49dac68d9/lib/rpush/daemon/apns2/delivery.rb#L18

Reply:
Hmm, I think having a partial fix is also welcome for others. I'm just not sure about the `client.on` stuff.

Have you seen #439 and https://github.com/ileitch/modis/issues/14 in case you are also using modis?

Reply:
Thanks for the feedback @aried3r 

I've dug into this issue a little more (I work with @rofreg) and I think I've identified the location of this memory leak (and also the memory leak associated with #439 (i don't believe modis is actually involved).  I'll write up my findings in #439 and push a solution to this PR.  

We can add the change to apns2 to this PR or open a new one if you prefer.

Reply:
Awesome, and thanks for your write-down! I'll try to have a look.



> We can add the change to apns2 to this PR or open a new one if you prefer.

Whatever is easier for you. If it's in individual commits, why not in this one.

Reply:
@aried3r, it was easier for me to split this PR into a couple PRs so I have opened #474, #475 and #476 

#474 and #475 are currently in our production instance; we don't use apns2 so I can't test #476.

I can collapse them back to this PR if you prefer but otherwise they are ready to be merged.

Reply:
Awesome! Makes it much easier to review and follow. I guess the new PRs supersede this PR?